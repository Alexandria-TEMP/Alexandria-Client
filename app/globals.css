@tailwind base;
@tailwind components;
@tailwind utilities;

@layer utilities {
  .text-balance {
    text-wrap: balance;
  }
}

@layer base {
  h1 {
    @apply text-2xl font-bold;
  }
  h2 {
    @apply text-xl font-semibold;
  }
  h3 {
    @apply text-lg;
  }
  h6 {
    @apply text-sm;
  }
}
