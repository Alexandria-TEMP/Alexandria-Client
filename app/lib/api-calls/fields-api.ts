import useSWR, { SWRResponse } from "swr";
import { ScientificFieldT } from "../types/api-types";
import { baseUrl, validateResponse } from "./api-common";

/**
 * Hook that fetches all scientific fields from the database, using SWR for caching, loading and error states
 * Fetching all tags from the database sounds like a very bad idea, but we would need some kind of lazy loading for MultiSelectAutocomplete otherwise
 * that is currently in todo
 * @returns the scientific field objects as array, as well as other SWR states (loading, error)
 */
export function useScientificFields(): SWRResponse<ScientificFieldT[], Error> {
  return useSWR(baseUrl + "/tags/scientific", async (...args) => {
    const response = await fetch(...args);
    await validateResponse(response);
    //disable reason: idk how to fix this and still get the correct type cause typescript
    // i have to look into this
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    const scientificFieldTags: ScientificFieldT[] = await response.json();
    return scientificFieldTags;
  });
}
