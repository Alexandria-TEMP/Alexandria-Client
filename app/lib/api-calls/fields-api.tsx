// methods in this file will be heavily changed once integration with back end it done
// just retrieve some dummy data for now

export async function getFields() {
  await new Promise((resolve) => setTimeout(resolve, 1000));
  const data = [
    {
      id: "1",
      tag: "Computer Science",
      tagType: "ScientificField",
    },
    {
      id: "2",
      tag: "Mathematics",
      tagType: "ScientificField",
    },
    {
      id: "3",
      tag: "Psychology & Psychiatry",
      tagType: "ScientificField",
    },
    {
      id: "4",
      tag: "Medicine",
      tagType: "ScientificField",
    },
  ];

  return data;
}
