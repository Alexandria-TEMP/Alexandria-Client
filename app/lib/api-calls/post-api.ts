import { PostT, PostCreationFormT, idT } from "../types/api-types";
import { validateResponse } from "./api-common";
import { baseUrl } from "./api-common";

/**
 * Gets data for a Post given their ID.
 * @async
 * @param id Post ID
 */
export default async function getPostData(id: idT): Promise<PostT> {
  // TODO
  await new Promise((resolve) => setTimeout(resolve, 100));

  return {
    title: "Post title",
    discussionIDs: [],
    renderStatus: "failure",
    collaboratorIDs: [1, 2], /// i think these should be different, but branch and post are strutured differently so we need both
    id: id,
    postType: "reflection",
    scientificFields: ["1", "2", "3"],
  };
}

/**
 * Method that sends a POST request to the server to create a new post, currently only metadata
 * @async
 * @param postCreationForm object containing post creation form data
 * @returns the newly created post
 */
export async function postPosts(
  postCreationForm: PostCreationFormT,
): Promise<PostT> {
  const jsonPost = JSON.stringify(postCreationForm);
  const response = await fetch(baseUrl + "/posts", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: jsonPost,
  });
  await validateResponse(response);
  //disable reason: idk how to fix this and still get the correct type cause typescript
  // i have to look into this
  // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
  const post: PostT = await response.json();
  return post;
}

/**
 * Method that sends a POST request to the server to upload a new file to an existing (non project) post
 * File technically gets sent as multipart form data
 * @async
 * @param postId the id of the post we want to upload files to
 * @param file the file object
 * @returns whether the request retuned a 200OK response
 */
export async function postPostsIdUpload(postId: idT, file: File) {
  const fileData = new FormData();
  fileData.append("file", file);

  const response = await fetch(baseUrl + "/posts/" + postId + "/upload", {
    method: "POST",
    body: fileData,
  });
  await validateResponse(response);
  return response.ok;
}
