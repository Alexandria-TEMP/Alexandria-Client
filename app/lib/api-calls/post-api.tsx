import { PostT } from "../api-types";

/**
 * Gets data for a Post given their ID.
 *
 * @async
 * @param id Post ID
 */
export default async function getPostData(id: string): Promise<PostT> {
  await new Promise((resolve) => setTimeout(resolve, 100));

  return {
    title: "Post title",
    status: "Open for review",
    authors: ["3"],
    contributors: ["1", "2"],
    collaborators: ["1", "2"], /// i think these should be different, but mr and post are strutured differently so we need both
    anonymous: false,
    createdAt: "10 May 2024",
    currentVersion: {
      id: "1",
      discussions: ["1", "1", "1", "1"],
    },
    id: id,
    postType: "Reflection",
    scientificFieldTags: ["1", "2", "3"],
    updatedAt: "11 May 2024",
    feedbackPreferences: "Community Discussion",
    completionStatus: "Ongoing",
  };
}
