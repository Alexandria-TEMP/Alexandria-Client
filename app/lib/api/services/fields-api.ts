import {
  ScientificFieldTagContainerT,
  ScientificFieldTagT,
  idT,
} from "../../types/api-types";
import { baseUrl, validateResponse } from "../api-common";

/**
 * TODO jsdoc when properly implemented
 */
export async function getFields() {
  // TODO
  await new Promise((resolve) => setTimeout(resolve, 1000));
  const data = [
    {
      id: "1",
      tag: "Computer Science",
      tagType: "ScientificField",
    },
    {
      id: "2",
      tag: "Mathematics",
      tagType: "ScientificField",
    },
    {
      id: "3",
      tag: "Psychology & Psychiatry",
      tagType: "ScientificField",
    },
    {
      id: "4",
      tag: "Medicine",
      tagType: "ScientificField",
    },
  ];

  return data;
}

/**
 * Fetches all scientific fields based on their IDs
 * @param ids scientific field IDs
 */
export async function fetchScientificFields(
  ids: idT[],
): Promise<ScientificFieldTagT[]> {
  let tags: ScientificFieldTagT[] = [];

  for (const id of ids) {
    const res = await fetch(`${baseUrl}/tags/scientific/${id}`);
    if (!res.ok) continue;
    const scientificField = (await res.json()) as ScientificFieldTagT;
    tags = [...tags, scientificField];
  }

  return tags;
}

/**
 * Fetch scientific field tag container
 * @param id of the container
 * @returns container object
 */
export async function fetchScientificFieldContainer(
  id: idT,
): Promise<ScientificFieldTagContainerT> {
  const res = await fetch(`${baseUrl}/tags/scientific/containers/${id}`);
  await validateResponse(res);
  const container = (await res.json()) as ScientificFieldTagContainerT;
  return container;
}

/**
 * Method that gets the scientific fiels from a container
 * @param id container id
 * @returns a list of scientific field tags in that container
 */
export async function fetchScientificFieldsFromContainer(
  id: idT,
): Promise<ScientificFieldTagT[]> {
  const container = await fetchScientificFieldContainer(id);
  return await fetchScientificFields(container.scientificFieldTagIDs);
}
