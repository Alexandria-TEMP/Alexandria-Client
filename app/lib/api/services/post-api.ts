import { PostT, ProjectPostT, idT } from "../../types/api-types";
import { PostUnionT, idPostUnionT } from "../../types/post-union";
import { baseUrl, validateResponse } from "../api-common";

/**
 * Fetches post or project post data in a unified object
 * @param id post or project post id
 * @returns post and optionally project post data
 */
export default async function fetchPostData(
  id: idPostUnionT,
): Promise<PostUnionT> {
  let projectPost = undefined;

  if (id.isProject) {
    const projectPostResponse = await fetch(
      `${baseUrl}/project-posts/${id.id}`,
    );
    await validateResponse(projectPostResponse);
    projectPost = (await projectPostResponse.json()) as ProjectPostT;
  }

  const postID = projectPost?.postID ?? (id.id as idT);
  const postResponse = await fetch(`${baseUrl}/posts/${postID}`);
  await validateResponse(postResponse);

  const post = (await postResponse.json()) as PostT;

  return { post, projectPost };
}
