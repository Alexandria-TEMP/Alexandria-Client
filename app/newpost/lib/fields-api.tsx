// methods in this file will be heavily changed once integration with back end it done
// just retrieve some dummy data for now

export function getFieldData(id: string) {
  // TODO
  if (id == "1") {
    return {
      id: "1",
      tag: "Computer Science",
      tagType: "ScientificField",
    };
  } else if (id == "2") {
    return {
      id: "2",
      tag: "Mathematics",
      tagType: "ScientificField",
    };
  } else if (id == "3") {
    return {
      id: "3",
      tag: "Psychology & Psychiatry",
      tagType: "ScientificField",
    };
  } else {
    return {
      id: "4",
      tag: "Medicine",
      tagType: "ScientificField",
    };
  }
}

export function getFields() {
  // TODO
  const data = [
    {
      id: "1",
      tag: "Computer Science",
      tagType: "ScientificField",
    },
    {
      id: "2",
      tag: "Mathematics",
      tagType: "ScientificField",
    },
    {
      id: "3",
      tag: "Psychology & Psychiatry",
      tagType: "ScientificField",
    },
    {
      id: "4",
      tag: "Medicine",
      tagType: "ScientificField",
    },
  ];

  return new Map(data.map((m) => [m.id, m]));
}
