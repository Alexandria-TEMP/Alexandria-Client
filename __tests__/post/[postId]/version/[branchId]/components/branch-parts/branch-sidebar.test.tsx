import BranchSidebar from "@/post/[postId]/(branch)/version/[branchId]/components/branch-parts/branch-sidebar";
import ChipList from "@/components/common/chip-list";
import AuthorCardList from "@/post/[postId]/components/cards/author-card-list";
import PostCardMini from "@/post/[postId]/components/cards/post-card-mini";
import { expect, describe, it } from "@jest/globals";
import { render } from "@testing-library/react";
import {
  fetchBranchData,
  fetchBranchUpdatedFieldsFallback,
} from "@/lib/api/services/branch-api";
import {
  dummyBranches,
  dummyPostUnion,
  dummyScientificFields,
} from "~/__tests__/__utils__/dummys";
import {
  fetchBranchCollaboratorsMemberIDs,
  fetchPostCollaboratorsMemberIDs,
} from "@/lib/api/services/collaborator-api";
import { fetchScientificFieldsFromContainer } from "@/lib/api/services/fields-api";
import { fetchPostData } from "@/lib/api/services/post-api";

jest.mock("@/lib/api/services/collaborator-api");
jest.mock("@/lib/api/services/fields-api");
jest.mock("@/lib/api/services/branch-api");
jest.mock("@/components/common/chip-list");
jest.mock("@/post/[postId]/components/cards/author-card-list");
jest.mock("@/post/[postId]/components/cards/post-card-mini");
jest.mock("@/lib/api/services/post-api");

describe("BranchSidebarTest", () => {
  (fetchBranchUpdatedFieldsFallback as jest.Mock).mockResolvedValue(
    dummyBranches["accepted"].updated,
  );
  (fetchBranchCollaboratorsMemberIDs as jest.Mock).mockResolvedValue([1]);
  (fetchPostCollaboratorsMemberIDs as jest.Mock).mockResolvedValue([1]);
  (fetchScientificFieldsFromContainer as jest.Mock).mockResolvedValue(
    dummyScientificFields,
  );
  (fetchBranchData as jest.Mock).mockResolvedValue(dummyBranches["accepted"]);
  (fetchPostData as jest.Mock).mockResolvedValue(dummyPostUnion.withProject);
  (ChipList as jest.Mock).mockReturnValue(<p>ChipList</p>);
  (AuthorCardList as jest.Mock).mockReturnValue(<p>AuthorCardList</p>);
  (PostCardMini as jest.Mock).mockReturnValue(<p>PostCardMini</p>);

  it("matches snapshot", async () => {
    const { container } = render(
      await BranchSidebar({ id: 1, isClosed: true }),
    );
    expect(container).toMatchSnapshot();
  });
});
