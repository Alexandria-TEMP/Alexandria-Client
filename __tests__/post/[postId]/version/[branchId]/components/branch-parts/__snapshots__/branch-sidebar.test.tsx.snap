// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BranchSidebarTest matches snapshot 1`] = `
<div>
  <div
    class="sticky top-20"
  >
    <h3>
      Version of
    </h3>
    <p>
      PostCardMini
    </p>
    <div
      class="h-4"
    />
    <h3>
      Contributors
    </h3>
    <p>
      AuthorCardList
    </p>
    <div
      class="h-4"
    />
  </div>
</div>
`;
